File: ch02/productinfo/java/server/src/main/java/ecommerce/ProductInfoImpl.java
package ecommerce;

import io.grpc.Status;
import io.grpc.StatusException;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class ProductInfoImpl extends ProductInfoGrpc.ProductInfoImplBase {

    private Map productMap = new HashMap<String, ProductInfoOuterClass.Product>();

    @Override
    public void addProduct(ProductInfoOuterClass.Product request,
                           io.grpc.stub.StreamObserver<ProductInfoOuterClass.ProductID> responseObserver) {
        UUID uuid = UUID.randomUUID();
        String randomUUIDString = uuid.toString();
        request = request.toBuilder().setId(randomUUIDString).build();
        productMap.put(randomUUIDString, request);
        ProductInfoOuterClass.ProductID id
                = ProductInfoOuterClass.ProductID.newBuilder().setValue(randomUUIDString).build();
        responseObserver.onNext(id);
        responseObserver.onCompleted();
    }

    @Override
    public void getProduct(ProductInfoOuterClass.ProductID request,
                           io.grpc.stub.StreamObserver<ProductInfoOuterClass.Product> responseObserver) {
        String id = request.getValue();
        if (productMap.containsKey(id)) {
            responseObserver.onNext((ProductInfoOuterClass.Product) productMap.get(id));
            responseObserver.onCompleted();
        } else {
            responseObserver.onError(new StatusException(Status.NOT_FOUND));
        }
    }
}

File: ch02/productinfo/cpp/server/ProductInfoImpl.cpp
//
// Created by lldx on 1/19/23.
//

#include "ProductInfoImpl.h"
#include <sstream>
grpc::Status ProductInfoImpl::addProduct(::grpc::ServerContext *context, const ::ecommerce::Product *request,
                                         ::ecommerce::ProductID *response) {
    std::cout << request->DebugString();
    std::string id = uuid_gen_test();
    (*server_info)[id]=*request;
    (*server_info)[id].set_id(id);
    response->set_value(id);
    return grpc::Status::OK;
}

grpc::Status ProductInfoImpl::getProduct(::grpc::ServerContext *context, const ::ecommerce::ProductID *request,
                                         ::ecommerce::Product *response) {
    std::cout << request->DebugString();
    if ( server_info->find(request->value()) != server_info->end() ){
        response->CopyFrom((*server_info)[request->value()]);
        return grpc::Status::OK;
    }
    return grpc::Status::CANCELLED;
}

ProductInfoImpl::~ProductInfoImpl() {

}

std::string ProductInfoImpl::uuid_gen_test() {
    static int i = 1;
    std::stringstream ss;
    ss << i ;
    i++;
    return ss.str();
}
